Generic comparison fails for conditional assignment

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->
**V doctor:**
```
V full version: V 0.4.1 52a055b.7169df3
OS: windows, Microsoft Windows 11 Home v22621 64-bit
Processor: 16 cpus, 64bit, little endian, 

getwd: C:\Development\Repositories\VProjects\scratch\generic_clip
vexe: C:\Development\Dependencies\v-main\v.exe
vexe mtime: 2023-09-21 04:50:16

vroot: OK, value: C:\Development\Dependencies\v-main
VMODULES: OK, value: C:\Development\Dependencies\vmodules
VTMP: OK, value: C:\Development\scratch\tmp

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2023.27-479-g7169df3e
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 e90c2620

```

**What did you do?**
`v -g -o vdbg cmd/v && vdbg .\src\main.v`
```v
module main

fn clip[N](x []N, min N, max N) []N {
	mut result := []N{cap: x.len}
	for value in x {
		result << if value < min {
			min
		} else if value > max {
			max
		} else {
			value
		}
	}
	return result
}

fn main() {
	a := [1, 2, 3, 4, 5]
	dump(a)
	dump(clip(a, 2, 4))
}

```

**What did you expect to see?**

The clipped version of the array: [2, 2, 3, 4, 4]


**What did you see instead?**
```
==================
C:/Development/scratch/tmp/main.3925996559115324338.tmp.c:6847: warning: implicit declaration of function 'tcc_backtrace'
C:/Development/scratch/tmp/main.3925996559115324338.tmp.c:12620: error: '{' expected (got ";")
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang

```
> [!IMPORTANT]
> You can vote for this issue using the 👍 reaction. More votes increase the issue's priority
> for developers.
>
> Take into account that only the 👍 reaction counts as a vote.
> Only reactions to the issue itself will be counted as votes, not comments.

--- a/vlib/v/checker/infix.v
+++ b/vlib/v/checker/infix.v
@@ -43,6 +43,21 @@ fn (mut c Checker) infix_expr(mut node ast.InfixExpr) ast.Type {
 	if node.op == .key_is {
 		c.inside_x_is_type = true
 	}
+	// `arr << if n > 0 { 10 } else { 11 }` set the right c.expected_type
+	if node.op == .left_shift && c.table.sym(left_type).kind == .array {
+		if mut node.right is ast.IfExpr {
+			if node.right.is_expr && node.right.branches.len > 0 {
+				mut last_stmt := node.right.branches[0].stmts.last()
+				if mut last_stmt is ast.ExprStmt {
+					expr_typ := c.expr(mut last_stmt.expr)
+					info := c.table.sym(left_type).array_info()
+					if expr_typ == info.elem_type {
+						c.expected_type = info.elem_type
+					}
+				}
+			}
+		}
+	}
 	mut right_type := c.expr(mut node.right)
 	if node.op == .key_is {
 		c.inside_x_is_type = false

--- a/vlib/v/tests/generic_comparison_for_conditional_assign_test.v
+++ b/vlib/v/tests/generic_comparison_for_conditional_assign_test.v
@@ -1,0 +1,21 @@
+fn clip[N](x []N, min N, max N) []N {
+	mut result := []N{cap: x.len}
+	for value in x {
+		result << if value < min {
+			min
+		} else if value > max {
+			max
+		} else {
+			value
+		}
+	}
+	return result
+}
+
+fn test_generic_comparison_for_conditional_assign() {
+	a := [1, 2, 3, 4, 5]
+	dump(a)
+	b := clip(a, 2, 4)
+	dump(b)
+	assert b == [2, 2, 3, 4, 4]
+}

